docs/topics/power-assert.md

```diff
@@ -0,0 +1,380 @@
+
+### Function call tracing
+
+The plugin supports function call tracing, which is similar to Rust's `dbg!` macro.
+Use it to trace and print function calls and their results:
+
+```kotlin
+class FunctionTrailingExampleTest {
+
+    fun exampleFunction(x: Int, y: Int): Int {
+        return x + y
+    }
+
+    @Test
+    fun testFunctionCallTracing() {
+        assert(exampleFunction(2, 3) + exampleFunction(1, 2) == 9)
+    }
+}
+```
+
+The output shows the intermediate results of functions calls:
+
+```text
+Assertion failed
+assert(exampleFunction(2, 3) + exampleFunction(1, 2) == 9)
+       |                     | |                     |
+       |                     | |                     false
+       |                     | 3
+       |                     | FunctionTrailingExampleTest@533bda92
+       |                     8
+       5
+       FunctionTrailingExampleTest@533bda92
+```
+
+### Soft assertions
+
+The Power-assert plugin supports soft assertions, which do not immediately fail the test but instead collect
+assertion failures and report them at the end of the test run.
+This can be useful when you want to see all assertion failures in a single run without stopping at the first failure.
+
+To enable soft assertions, implement the way you will collect error messages:
+
+```kotlin
+fun <R> assertSoftly(block: AssertScope.() -> R): R {
+    val scope = AssertScopeImpl()
+    val result = scope.block()
+    if (scope.errors.isNotEmpty()) {
+        throw AssertionError(scope.errors.joinToString("\n"))
+    }
+    return result
+}
+
+interface AssertScope {
+    fun assert(assertion: Boolean, message: (() -> String)? = null)
+}
+
+class AssertScopeImpl : AssertScope {
+    val errors = mutableListOf<String>()
+    override fun assert(assertion: Boolean, message: (() -> String)?) {
+        if (!assertion) {
+            errors.add(message?.invoke() ?: "Assertion failed")
+        }
+    }
+}
+```
+
+After that, you could use it in your test code:
+
+```kotlin
+// Import the assertSoftly() function
+import org.example.assertSoftly
+        
+class SoftAssertExampleTest1 {
+
+    data class Employee(val name: String, val age: Int, val salary: Int)
+
+    @Test
+    fun `test employees data`() {
+        val employees = listOf(
+            Employee("Alice", 30, 60000),
+            Employee("Bob", 45, 80000),
+            Employee("Charlie", 55, 40000),
+            Employee("Dave", 150, 70000)
+        )
+
+        assertSoftly {
+            for (employee in employees) {
+                assert(employee.age < 100) { "${employee.name} has an invalid age: ${employee.age}" }
+                assert(employee.salary > 50000) { "${employee.name} has an invalid salary: ${employee.salary}" }
+            }
+        }
+    }
+}
+```
+
+In the output, all the `assert()` function error messages will be printed one after another:
+
+```text
+Charlie has an invalid salary: 40000
+Dave has an invalid age: 150
```

Comment:
This is not the output I would have expected if power-assert was enabled for the `AssertScope.assert` function. I would expect something more along the lines of the following, if the function was configured correctly.

```text
Charlie has an invalid salary: 40000
assert(employee.salary > 50000) { "${employee.name} has an invalid salary: ${employee.salary}" }
       |        |
       |        40000
       Employee(name="Charlie", age=55, salary=40000)
Dave has an invalid age: 150
assert(employee.age < 100) { "${employee.name} has an invalid age: ${employee.age}" }
       |        |
       |        150
       Employee(name="Dave", age=150, salary=70000)
```

I think configuring the function is just missing, and you need something like the following to make this work as expected.

```kotlin
powerAssert {
    functions = listOf("[package AssertScope is from].AssertScope.assert")
}
```

Reply:
Exactly, I forgot to add both `"org.example.AssertScope.assert", "org.example.assertSoftly"`functions

Reply:
`org.example.assertSoftly` is not needed and will result in a compiler warning, only `org.example.AssertScope.assert` is needed.